### PR TITLE
Webui: add flask implementation #4041

### PR DESCRIPTION
**Pull request by tbeerman**

### CODE REVIEW

File: lib/rucio/web/ui/flask/common/utils.py [hunk 1]
```diff
@@ -0,0 +1,626 @@
...
+# - Mario Lassnig <mario.lassnig@cern.ch>, 2014-2018
+# - Vincent Garonne <vincent.garonne@cern.ch>, 2015
+# - Martin Barisits <martin.barisits@cern.ch>, 2016-2020
+# - Ruturaj Gujar <ruturaj.gujar23@gmail.com>, 2019
+# - Jaroslav Guenther <jaroslav.guenther@cern.ch>, 2019-2020
+# - Eli Chadwick <eli.chadwick@stfc.ac.uk>, 2020
+# - Benedikt Ziemons <benedikt.ziemons@cern.ch>, 2020
+# - Patrick Austin <patrick.austin@stfc.ac.uk>, 2020
+#
+# PY3K COMPATIBLE
```
Comment (bziemons) [inline lib/rucio/web/ui/flask/common/utils.py]: this may be removed

File: lib/rucio/web/ui/flask/__init__.py [hunk 2]
```diff
@@ -0,0 +1,5 @@
+# Copyright European Organization for Nuclear Research (CERN)
```
Comment (bziemons) [inline lib/rucio/web/ui/flask/__init__.py:1]: why is this copyright different than the others?

File: lib/rucio/web/ui/flask/common/utils.py [hunk 3]
```diff
@@ -0,0 +1,626 @@
...
+            return render_template('problem.html', msg='Cannot get auth token. It is possible that the presented identity %s is not mapped to any Rucio account %s at VO %s.') % (html_escape(username), html_escape(ui_account), html_escape(ui_vo))
+        else:
+            return render_template('problem.html', msg='Cannot get auth token. It is possible that the presented identity %s is not mapped to any Rucio account %s.') % (html_escape(username), html_escape(ui_account))
+
+    return finalize_auth(token, 'userpass')
+
+
+def saml_auth(method, data=None):
+    """
+    # TODO: Validate for Flask
```
Comment (bziemons) [inline lib/rucio/web/ui/flask/common/utils.py:431]: I guess this should be done before the merge

File: lib/rucio/web/ui/flask/common/utils.py [hunk 4]
```diff
@@ -0,0 +1,626 @@
...
+            return finalize_auth(token, 'saml', cookie_extra)
+
+        return render_template("problem.html", msg="Not authenticated")
+
+    return render_template("problem.html", msg="Error while processing SAML")
+
+
+def oidc_auth(account, issuer, ui_vo=None):
+    """
+    # TODO: Validate for Flask
```
Comment (bziemons) [inline lib/rucio/web/ui/flask/common/utils.py:553]: I guess this should be done before the merge

File: lib/rucio/web/ui/flask/templates/select_login_method.html [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  <h4 class="card-title text-center mb-4 mt-1"><img src="/media/RucioUI.png" width="450px"></h4>
+  <hr>
+  <p class="text-success text-center">Choose Login Method</p>
+  <p>
+    <a href="" class="btn btn-block btn-outline-primary btn-lg" role="button" onclick="get_link(this, 'x509')" height="auto"><i class="fab fa-2x icon-certificate vertical-align-middle"></i><span class="inline-block padding-bottom-2">   X509 Certificate</span></a>
+    <a href="" class="btn btn-block btn-outline-rucio btn-lg" role="button" onclick="get_link(this, 'login')" height="auto"> <i class="fab fa-2x icon-rucio vertical-align-middle"></i><span class="inline-block padding-bottom-2">   Rucio Userpass</span></a>
+    {% if saml_support %}
+        <a href="" class="btn btn-block btn-outline-cern btn-lg" role="button" onclick="get_link(this, 'saml')" height="auto"> <i class="fab fa-2x icon-cern vertical-align-middle"></i><span class="inline-block padding-bottom-2">   CERN SSO Log-In</span></a>
+    {% endif %}
+    {# TODO: Fix OIDC buttons #}
```
Comment (bziemons) [inline lib/rucio/web/ui/flask/templates/select_login_method.html:55]: is this still a todo?

File: lib/rucio/web/ui/flask/templates/select_login_method.html [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+  </p>
+  <form>
+  <div class="form-group">
+    <div class="input-group">
+      <div class="input-group-prepend">
+        <span class="input-group-text" style="width: 2.75em"> <i class="fa fa-user"></i> </span>
+      </div>
+      <input name="accountname" class="form-control" placeholder="Optionally specify Rucio account name ..." type="accountname" id="accountname" value="">
+    </div> <!-- input-group.// -->
+    {# TODO: Fix multi-VO buttons #}
```
Comment (bziemons) [inline lib/rucio/web/ui/flask/templates/select_login_method.html:69]: is this still a todo?

File: lib/rucio/web/ui/flask/common/utils.py [hunk 7]
```diff
@@ -0,0 +1,626 @@
...
+COOKIES = []
+
+
+def html_escape(s, quote=True):
+    return escapefunc(s, quote)
+
+
+def prepare_saml_request():
+    """
+    TODO: Validate for Flask
```
Comment (bziemons) [inline lib/rucio/web/ui/flask/common/utils.py:105]: I guess this should be done before the merge

File: lib/rucio/web/ui/flask/main.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+from rucio.common.config import config_get, config_get_bool
+from rucio.web.rest.utils import generate_http_error_flask
+from rucio.web.ui.flask.common.utils import get_token, authenticate, userpass_auth, x509token_auth, saml_auth, oidc_auth, finalize_auth, AUTH_ISSUERS, SAML_SUPPORT
+
+MULTI_VO = config_get_bool('common', 'multi_vo', raise_exception=False, default=False)
+POLICY = config_get('policy', 'permission')
+ATLAS_URLS = ()
+OTHER_URLS = ()
+
+application = Flask(__name__)
```
Comment (bziemons) [inline lib/rucio/web/ui/flask/main.py:26]: since all the Flask endpoints are currently in one application in `lib/rucio/web/rest/flaskapi/v1/main.py`, it should be possible to register the ui as a blueprint there as well.

Actually registering it depends on #4043

File: lib/rucio/web/ui/static/subscriptions.js [hunk 9]
```diff
@@ -135,7 +135,7 @@ $(document).ready(function(){
         insertParam('account', chosen_account);
     }
 
-    var link_list_rules = '/list_rules?account=' + chosen_account;
+    var link_list_rules = '/r2d2?account=' + chosen_account;
```
Comment (bziemons) [inline lib/rucio/web/ui/static/subscriptions.js:138]: unrelated change?

**Review by bziemons — COMMENTED** [review]
I would really like to see the ui as a Flask blueprint. The rest of the comments are just things that caught my attention.

Maybe for the copyright in the .js files we could add the format to the add_header script, because those seem to be outdated.

File: lib/rucio/web/ui/flask/__init__.py (hunk 2 above)
Comment (tbeerman) [reply lib/rucio/web/ui/flask/__init__.py:1]: No idea, didn't even look at it. Just copied it from the main ui folder.

File: lib/rucio/web/ui/flask/common/utils.py (hunk 3 above)
Comment (tbeerman) [reply lib/rucio/web/ui/flask/common/utils.py:431]: I don't really have a setup to this. It was developed by an external who is not in the project anymore. This is not deployed anywhere, yet.

File: lib/rucio/web/ui/flask/common/utils.py (hunk 4 above)
Comment (tbeerman) [reply lib/rucio/web/ui/flask/common/utils.py:553]: Same as above.

File: lib/rucio/web/ui/flask/templates/select_login_method.html (hunk 5 above)
Comment (tbeerman) [reply lib/rucio/web/ui/flask/templates/select_login_method.html:55]: Yes, that's why I put it there.

File: lib/rucio/web/ui/flask/templates/select_login_method.html (hunk 6 above)
Comment (tbeerman) [reply lib/rucio/web/ui/flask/templates/select_login_method.html:69]: Yes, that's why I put it there.

File: lib/rucio/web/ui/flask/common/utils.py (hunk 7 above)
Comment (tbeerman) [reply lib/rucio/web/ui/flask/common/utils.py:105]: See previous comments.

File: lib/rucio/web/ui/flask/main.py (hunk 8 above)
Comment (tbeerman) [reply lib/rucio/web/ui/flask/main.py:26]: What would be the advantage?

File: lib/rucio/web/ui/static/subscriptions.js (hunk 9 above)
Comment (tbeerman) [reply lib/rucio/web/ui/static/subscriptions.js:138]: Not really. In the web.py ui, list_rules is just a redirect to r2d2. This saves me the endpoint in the flask app and it won't break web.py.

File: lib/rucio/web/ui/flask/main.py (hunk 8 above)
Comment (bziemons) [reply lib/rucio/web/ui/flask/main.py:26]: getting rid of multiple WSGI script aliases in the apache config. #4043 is coming anyway, so having to configure the ui endpoint specifically in the apache config just makes it harder and more confusing to set up.

Comment (tbeerman) [reply lib/rucio/web/ui/flask/main.py:26]: okay. In the webui case it's anyway only one alias pointing to '/' and it's usually not deployed together with the api. But I'll add it to be consistent with the api.

Comment (bziemons) [reply lib/rucio/web/ui/flask/main.py:26]: ah, right, I was thinking of having the /ui endpoint for testing/demo purposes. Is that not supported?

Comment (tbeerman) [reply lib/rucio/web/ui/flask/main.py:26]: yes, that's the only real use case I can think of where you would have the rest api and webui running within the same apache. Like it's done in the dev container.

**Comment by tbeerman** [comment]
I've integrated the changes that I could at the moment. The OIDC and SAML support is for another PR. It's nowhere used in production, yet, and even the webpy implementation needs some testing.

**Review by bziemons — APPROVED** [review]
Blueprint looks good. Thanks
